<h1 align=center><font size = 6> News Classification. </font></h1>

# Author : Akshay Godse.
#### Data Science Intern

## Problem Statement :

To build a model to accurately classify a piece of news as REAL or FAKE.



Using sklearn,  build a TfidfVectorizer on the provided dataset. Then, initialize a PassiveAggressive Classifier and fit the model. In the end, the accuracy score and the confusion matrix tell us how well our model fares.

### Import all the required libraries.

In [36]:
import numpy as np       # library that handels data in vectorized manner.
import pandas as pd      # libary for data analysis.


### Machine learning Libraries ####
from sklearn.model_selection import train_test_split           # library that splits the dataset into training set and testing set.
from sklearn.feature_extraction.text import TfidfVectorizer    # library for Tdfvectorizer.
from sklearn.linear_model import PassiveAggressiveClassifier   # library for PassiveAggressiveClassifie.
from sklearn.metrics import accuracy_score, confusion_matrix   # library for accuracy score calculation and confusion matrics.

import itertools   # library that creates the iterators.

import matplotlib as plt

print("All the required libraries are imported....!")

All the required libraries are imported....!


### Read CSV file dataset into the pandas data frame.

In [19]:
df_news = pd.read_csv("news.csv")

#now lets see how our data frame looks like.
df_news.head()

Unnamed: 0                                              title  \
0        8476                       You Can Smell Hillary’s Fear   
1       10294  Watch The Exact Moment Paul Ryan Committed Pol...   
2        3608        Kerry to go to Paris in gesture of sympathy   
3       10142  Bernie supporters on Twitter erupt in anger ag...   
4         875   The Battle of New York: Why This Primary Matters   

                                                text label  
0  Daniel Greenfield, a Shillman Journalism Fello...  FAKE  
1  Google Pinterest Digg Linkedin Reddit Stumbleu...  FAKE  
2  U.S. Secretary of State John F. Kerry said Mon...  REAL  
3  — Kaydee King (@KaydeeKing) November 9, 2016 T...  FAKE  
4  It's primary day in New York and front-runners...  REAL

### Initial Data Analysis and Data Prepration.

In [20]:
# Get the shape of the data frame.
df_news.shape

(6335, 4)

In [22]:
#rename the column name Unnamed: 0 to id.
df_news.rename(columns={ 'Unnamed: 0' : 'id'}, inplace=True)

#lets have a look into the data frame.
df_news.head()

id                                              title  \
0   8476                       You Can Smell Hillary’s Fear   
1  10294  Watch The Exact Moment Paul Ryan Committed Pol...   
2   3608        Kerry to go to Paris in gesture of sympathy   
3  10142  Bernie supporters on Twitter erupt in anger ag...   
4    875   The Battle of New York: Why This Primary Matters   

                                                text label  
0  Daniel Greenfield, a Shillman Journalism Fello...  FAKE  
1  Google Pinterest Digg Linkedin Reddit Stumbleu...  FAKE  
2  U.S. Secretary of State John F. Kerry said Mon...  REAL  
3  — Kaydee King (@KaydeeKing) November 9, 2016 T...  FAKE  
4  It's primary day in New York and front-runners...  REAL

In [23]:
# lets remove the blank spaces and Nan values in the data set.
nan_value = float("NaN")
df_news.replace("", nan_value, inplace=True)
df_news.dropna(subset = ["id"], inplace=True)
print("Drop Sucessfull")
#now look at the shape of our data frame.
print("Size of the data frame is :", df_news.shape)

Drop Sucessfull
Size of the data frame is : (6335, 4)


In [24]:
# Get the labels into the another data frame called df_labels.
df_labels = df_news.label

#look at the df_labels.
df_labels.head()

0    FAKE
1    FAKE
2    REAL
3    FAKE
4    REAL
Name: label, dtype: object

### Split the datasets into training and testing data.
here we are splitting data 70 % into training set and 30 % into testing set by using random state 8 to achieve the highest accuracy.

In [28]:
X_train, X_test, y_train, y_test = train_test_split(df_news['text'], df_labels, test_size = 0.3, random_state = 8)

print("Data Splitted successfully into training and testing set. ")

Data Splitted successfully into training and testing set. 


### Initialize Tdfvectorizer and fit the data.

In [30]:
# initializing tdf vectorizer.
Tdfv = TfidfVectorizer(stop_words='english', max_df=0.7)

#now lets fit and transform data of the trainning set.
Tdf_train = Tdfv.fit_transform(X_train)
#now transform the data of testing set.
Tdf_test = Tdfv.transform(X_test)

### Now Initialize the PassiveAggressiveClassifier.

In [39]:
PA_Classifier = PassiveAggressiveClassifier(max_iter = 50)

#now lets fir the data.
PA_Classifier.fit(Tdf_train,y_train)


PassiveAggressiveClassifier(C=1.0, average=False, class_weight=None,
                            early_stopping=False, fit_intercept=True,
                            loss='hinge', max_iter=50, n_iter_no_change=5,
                            n_jobs=None, random_state=None, shuffle=True,
                            tol=0.001, validation_fraction=0.1, verbose=0,
                            warm_start=False)

### Now predict the result and calculate the accuracy.

In [41]:
# Predict the results.
y_pred = PA_Classifier.predict(Tdf_test)
print("Predicted Result is :")
y_pred

Predicted Result is :


array(['REAL', 'REAL', 'REAL', ..., 'REAL', 'FAKE', 'FAKE'], dtype='<U4')

In [44]:
# Calculate the accuracy of our model.
Accuracy_score = accuracy_score(y_test,y_pred)
print(f"Accuracy Score is : {round(Accuracy_score * 100,2)} %")

Accuracy Score is : 93.53 %


Now look here our model achieved 93.53 % accuracy, it shows that our maodel best fitted.

### Now lets build the confusion matrix for the model.

In [46]:
C_matrix = confusion_matrix(y_test,y_pred, labels=['FAKE','REAL'])
C_matrix

array([[898,  49],
       [ 74, 880]], dtype=int64)

Here the accuracy score and the confusion matrix shows that our model well fitted into the data set and predicted the results by achieving the highest accuracy.

Thank You..!<br>
AKSHAY GODSE,<br>
Cloud Counselage Pvt Ltd.